Electric storage controls and clean up

EnergyArchmage opened the pull request:
This is a project to rework the electric load center for grid connection and to fix numerous bugs.  Users had requested more features for controlling electric storage.  A new power conversion device for AC to DC is added for grid charging DC storage. 

This project addressed the following issues. 
 Issue #3211 -- capacity limited to on-site generator charging.
 Issue #5004 -- battery won't charge at its max rate
Issue #4273 -- documentation of storage rules and behavior. 
Issue #5299 -- Warn if thermal to electric power ratio not set for FollowThermal\* generator operation schemes
Issue #5302, detect and warn if multiple load centers and one battery (appears to work with simple storage).
Issue #5301, fatal kinetic battery with no information, was hard crash in  v8.1. 
Issue #4921, false object not found error, units in kinetic battery differ from simple storage.
Issue #4639, missing ElectricLoadCenter:Distribution object should lead to error message.
Issue #4113, RDD and documentation don't match for whole-building electric reports
Issue #5303, Sign reversed in kinetic battery model Electric Storage Production Decrement Ener

The flow is a little different and some output variables and meters get created in different orders so there are diffs in RDD, AUD, and MTD that are expect. 

Changes how photovoltaics and inverter are metered. 

Code base completely reworked to be more OO.

EnergyArchmage committed:
Initial OO refactor of electric service code

time to see how CI likes this set of code.  removed legacy Electric
power manager and replace with new OO version.  Attempting to replicate
old behavior at this point

EnergyArchmage committed:
fix build errors and warnings

get rid of std::make_unique, not in all versions of CI machines.
try some warning fixes

EnergyArchmage committed:
Fix unit tests

finish out revising new unit tests.  remove make_unique.

EnergyArchmage committed:
More build errors and warnings

change for loop to use std::size_t, missed one make_unique, rearrange
variables

EnergyArchmage committed:
transformer fixes

EnergyArchmage committed:
Fix tests

have both generator num and plant component num, revert sign correction
to match legacy code, sum up number of storage devices

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -78133,21 +78196,169 @@ ElectricLoadCenter:Distribution,
        \key DirectCurrentWithInverterACStorage
        \default AlternatingCurrent
-  A7,  \field Inverter Object Name
+!  A ,  \field Inverter Object Type
```
The commented lines should be removed before merging into develop.  I realize they are there for future use.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -78133,21 +78196,169 @@ ElectricLoadCenter:Distribution,
+       \note A single bidirectional device will reference both an inverter object (DC to AC) and a converter object (AC to DC).
+       \type object-list
+       \object-list ConverterList
+  
```
Stray blank line.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -78133,21 +78196,169 @@ ElectricLoadCenter:Distribution,
+       \type real
+       \units W
+  A11, \field Storage Charge Grid Supply Power Fraction Schedule Name
+       \note controls timing and magnitude of charging from grid
```
Nit-picking - captialize "Controls"

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -78133,21 +78196,169 @@ ElectricLoadCenter:Distribution,
+       \type object-list
+       \object-list ScheduleNames
+  A12, \field Storage Converter Object Name
+       \note name of ElectricLoadCenter:Storage:Converter used to convert AC to DC when charging DC storage from grid supply.
```
Captialize "Name" . . . 
There are more notes below where the first line needs an inital cap.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -78133,21 +78196,169 @@ ElectricLoadCenter:Distribution,
+       \note Select method to govern how storage charge and discharge is controlled
+       \type choice
+       \key DemandLimit
+       \note FacilityDemandLimit indicates that storage discharge control will limit facility power demand drawn from the utility service 
```
Is this key DemandLimit or FacilityDemandLimit

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -78133,21 +78196,169 @@ ElectricLoadCenter:Distribution,
+       \note if omitted a schedule value of 1.0 is used. 
+       \type object-list
+       \object-list ScheduleNames
+  A15, \field Storage Discharge Track Schedule Name
```
Would you be open to changing this to a fraction schedule and adding a field for the max/nominal value?  Seems that experimenting with storage size would require a new schedule here which would be tedious.

EnergyArchmage replied:
For this PR, the IDD has been reverted.  Will incorporate these comments moving forward

mjwitte commented on src/EnergyPlus/ElectricPowerServiceManager.hh:
```diff
@@ -0,0 +1,1036 @@
+	DCtoACInverter( DCtoACInverter const & ) = default;
+
+	// Move Constructor
+#if !defined(_MSC_VER) || defined(__INTEL_COMPILER) || (_MSC_VER>=1900)
```
What is this for?  Just trying to understand - I don't see this over in #5426

EnergyArchmage replied:
This is some C++ stuff from @DeadParrot which I cannot explain.   I think it has to do with not being able to copy unique pointers and the default move constructor getting something wrong. Because these class are instantiated as unique pointers this stuff is needed here.

DeadParrot replied:
[Visual C++ 2013 doesn't support marking move constructors (or move assignment) as `= default`](https://msdn.microsoft.com/en-us/library/dn457344.aspx). This is fixed in VC++ 2015. It is not related to unique_ptr.

The reason for having them declared `= default` here is to make them private -- if you leave that out they are public. As to why this class has private construction I'm not sure: if it is meant to be a base class the constructors should be protected, not private, and the destructors should be virtual and assignment operators would normally be protected. If this is for a Named Constructor usage (unlikely) then where are the named constructors. I think some C++ review is needed here once the intent is clear.

In the sample AirTerminalUnit code I provided to @EnergyArchmage this `= default` usage was found in a protected section of a proper base class, which is not what we have here.

mjwitte commented on src/EnergyPlus/ElectricPowerServiceManager.hh:
```diff
@@ -0,0 +1,1036 @@
+
+namespace EnergyPlus {
+
+namespace ElectricPowerService {
```
Maybe I'm wrong, but don't we try to keep the namespace and the filename consistent?  ElectricPowerService vs ElectricPowerServiceManager

EnergyArchmage committed:
revert IDD

IDD changes don't really need to go with this

EnergyArchmage committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into Electric-Service-Refactor

EnergyArchmage committed:
new features, more refactor

let use see what CI says.

EnergyArchmage committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into Electric-Service-Refactor

# Conflicts:
#	src/Transition/OutputRulesFiles/OutputChanges8-4-0-to-8-5-0.md

EnergyArchmage committed:
Collection of fixes, address issues, docs

clean up. basic I/O ref content

EnergyArchmage committed:
fix build warnings, clean up

EnergyArchmage committed:
state of charge control limits for kinetic battery

EnergyArchmage committed:
various corrections, docs, transition.

change testfiles to match new output variable split.  update output vars
in I/O ref.

EnergyArchmage committed:
fix a unit test, range based fors

EnergyArchmage committed:
Fix and expand unit tests

EnergyArchmage committed:
fix RDD diffs, more range based fors,

don't create Electric Load Center output variables.

EnergyArchmage committed:
fix special chars in eng ref,

revise load center updates for when no generators

EnergyArchmage committed:
eng ref special chars

check the box this time

EnergyArchmage committed:
add integration tests, overview diagram,

also surplus rate output variable.

EnergyArchmage committed:
EMS integration test file, input rules

add rule for changed internal EMS variable, mirrors change to output
variable.

EnergyArchmage committed:
fix unit tests assert

more range based fors,, detect if gen list used on more than one load
center

EnergyArchmage committed:
I/O Ref improvements

add 5 diagrams for buss type showing storage operation control volume.

EnergyArchmage committed:
fix converter reinit at begin env

EnergyArchmage commented:
I believe this is good to go.  Do not expect any input/output changes.  The four integration test files (Generators_Transformer.idf, ShopWithPVandBattery.idf, ShopWithPVandStorage,idf, and ShopWithSimplePVT) are expected to show diffs.  Diffs are expected because 1) program changed to allow rates of charging and charging that are higher than generators (issues 3211 and 5004), and 2) the DC power from PV no longer goes thru the inverter before storing in DC storage.  

I believe this pull request address the following issues

```
1. Issue #3211 -- capacity limited to on-site generator charging.
2. issue #5004 -- battery won't charge at its max rate
3. Issue #4273 -- documentation of storage rules and behavior. 

5. Issue #5299 -- Warn if thermal to electric power ratio not set for FollowThermal* generator operation schemes
6. Issue #5302, detect and warn if multiple load centers and one battery (appears to work with simple storage).
7. Issue #5301, fatal kinetic battery with no information, was hard crash in  v8.1. 
8. issue #4921, false object not found error, units in kinetic battery differ from simple storage.
9. Issue #4639, missing ElectricLoadCenter:Distribution object should lead to error message.
10. Issue #4113, RDD and documentation don't match for whole-building electric reports
11. Issue #5303, Sign reversed in kinetic battery model Electric Storage Production Decrement Energy    
```

Not sure what is going on with ubuntu integration coverage test failure integration.ShopWithPVandBattery, how can see what happened there?

Myoldmopar commented:
Looks like a crash...

EnergyArchmage committed:
clear out conflicting binary

Myoldmopar commented:
I guess this branch needs Transition source, right?  That can wait until after the I/O freeze if you can figure out the ShopWithPVandBattery problem.  Does it run successfully for you locally?  I'm building also...

Myoldmopar commented:
On a Mac debug build, it runs fine.  I'm not sure if it was a fluke or what at this point.  Ugh.  This is a **ton** of important improvements and this should go in.  I'm almost ready to just let it go in, understanding that develop might have that one failure........  I can spin up an Ubuntu debug build, but it'll be a little while before I can reproduce or not.

Can you remedy the conflict by copying/pasting the rows in the Rules8...xls file from develop and this branch?

EnergyArchmage commented:
so far Rules8..xls has completely buggered up my github desktop, I can't merge or clear that binary conflict at all.  I need to purge my drive I guess. 

Yes, ShopWithPVandBattery runs fine here, I have no info about the ubuntu crash.

Myoldmopar commented:
Ubuntu is building now, I'll let you know.  

As for the Rules file, the easiest way to resolve the binary conflict is probably to:
- Browse to the file in Github in the develop branch, and download the [raw version](https://github.com/NREL/EnergyPlus/blob/develop/src/Transition/InputRulesFiles/Rules8-4-0-to-8-5-0.xls?raw=true)
- Browse to the file in this Electric branch, and download that raw version.  Uh oh.  I think you deleted it entirely from your branch.  OK, have to go back to a prior commit and download that [raw version](https://github.com/NREL/EnergyPlus/blob/fd419f1b929ca0ef2f9c4ebb65623d6f6d6d7a38/src/Transition/InputRulesFiles/Rules8-4-0-to-8-5-0.xls?raw=true).  
- So then just open both of those workbooks and make sure your new rules are in there along with whatever was added in develop.  
- Add that file back into your branch in the right location, commit it, and push it.

Myoldmopar commented:
Floating point exception in ElectricPowerServiceManager.cc:3171: 

```
Program received signal SIGFPE, Arithmetic exception.
0x00007ffff5491f2d in EnergyPlus::ElectricStorage::simulateKineticBatteryModel (this=0x1278930, powerCharge=@0x1277f08: 36769.115635336995, powerDischarge=@0x1277f10: 0, charging=@0x1277f18: true, 
    discharging=@0x1277f19: false, controlSOCMaxFracLimit=0.94999999999999996, controlSOCMinFracLimit=0.20000000000000001) at /home/elee/eplus/EnergyPlus/src/EnergyPlus/ElectricPowerServiceManager.cc:3171
3171                Real64 Tnew = qmaxf / std::abs( Inew );
```

EnergyArchmage committed:
add rules back

DeadParrot commented:
This needs a C++ review: private default constuctors, passing strings and vectors by value, ...

amirroth commented:
Tsk tsk tsk

On Fri, Feb 19, 2016 at 4:43 PM, Stuart Mentzer notifications@github.com
wrote:

> This needs a C++ review: private default constuctors, passing strings and
> vectors by value, ...
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/NREL/EnergyPlus/pull/5437#issuecomment-186420595.

EnergyArchmage committed:
trap div by zero

mbadams5 commented:
@DeadParrot why do you think private default constructors are bad, especially if we are using a static factory method to instantiate the object? Are you wanting a protected default constructor instead so this can potentially be inherited in the future?

DeadParrot commented:
@mbadams5 Private default constructors are just unusual so seem worth a look. If the factory can create them and the string arg constructors are public why does the default constructor need to be private? Putting them and copy and move constructors in private says that it is not OK for general code to instantiate them and I didn't see a reason for it here. I just took a quick glance so maybe this all makes sense but it would be one thing as a reviewer I'd want to see explained.

Also, why the string arg constructors don't use member initialization (but the default constructors do). Just seems like this needs a careful review of C++ usage.

Myoldmopar committed:
Resolve conflict with rules file

Myoldmopar commented:
@EnergyArchmage I remedied the transition rules file, so this doesn't have that conflict anymore.  I also re-ran the Ubuntu debug build:  

```
Program received signal SIGFPE, Arithmetic exception.
0x00007ffff54921b3 in EnergyPlus::ElectricStorage::simulateKineticBatteryModel (this=0x1278930, powerCharge=@0x1277f08: 36769.115635336995, powerDischarge=@0x1277f10: 0, charging=@0x1277f18: true, 
    discharging=@0x1277f19: false, controlSOCMaxFracLimit=0.94999999999999996, controlSOCMinFracLimit=0.20000000000000001) at /home/elee/eplus/EnergyPlus/src/EnergyPlus/ElectricPowerServiceManager.cc:3183
3183                    qmaxf = qmax * k * c * T0 / ( 1.0 - std::exp( -k * T0 ) + c * ( k * T0 - 1.0 + std::exp( -k * T0 ) ) );
(gdb) print qmax
$1 = 86.099999999999994
(gdb) print k
$2 = 0.58740000000000003
(gdb) print c
$3 = 0.37
(gdb) print T0
$4 = 0
```

EnergyArchmage committed:
fix kinetic battery

return early once storage limits are hit.

EnergyArchmage committed:
remerge output variable change log

Myoldmopar commented:
9442c20 fixes the integration file issue.  We are merging this in.

@DeadParrot @mjwitte @EnergyArchmage I am adamant on getting this in because it fixes _so_ many important bugs, and because we are lucky to have an extremely small window to get this in this morning.  I will open up a high priority, high severity ticket to go back over this code and do a more thorough review of the C++ to get @DeadParrot's comments addressed.  But this is going in now to get the IDD changes in place.
